iocage_lib/ioc_exec.py

```diff
@@ -149,22 +151,33 @@ def exec_jail(self):
 
                 if self.msg_return or self.msg_err_return:
                     p = su.Popen(cmd, stdout=su.PIPE, stderr=su.PIPE,
-                                 close_fds=True, bufsize=1)
-
+                                 close_fds=True, bufsize=0)
+
+                    # Courtesy of @william-gr
+                    # service(8) and some rc.d scripts have the bad habit of
+                    # exec'ing and never closing stdout/stderr. This makes
+                    # sure we read only enough until the command exits and do
+                    # not wait on the pipe to close on the other end.
+                    #
+                    # Same issue can be demonstrated with:
+                    # $ jexec 1 service postgresql onerestart | cat
+                    # ... <hangs>
+                    # postgresql rc.d command never closes the pipe
                     rtrn_stdout = b''
                     rtrn_stderr = b''
+                    for i in ('stdout', 'stderr'):
+                        fileno = getattr(p, i).fileno()
+                        fl = fcntl.fcntl(fileno, fcntl.F_GETFL)
+                        fcntl.fcntl(fileno, fcntl.F_SETFL, fl | os.O_NONBLOCK)
 
-                    while True:
+                    while p.poll() is None:
                         r = select.select([p.stdout.fileno(),
-                                           p.stderr.fileno()], [], [], 0.5)[0]
+                                           p.stderr.fileno()], [], [], 0)[0]
```

Comment:
0 on every loop in make the CPU spin, my suggestion was to change timeout to 0 only if `p` has finished

at the end of loop:

```
if p.poll() is not None:
    if timeout == 0:
        break
    else:
        timeout = 0
```